// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. mapper.proto  
syntax = "proto3";

package mapper;

service Mapper {
    rpc Map(MapRequest) returns (MapResponse);
    rpc GetPairs(GetPairsRequest) returns (GetPairsResponse);
}

// sends centroids and the shard(repeated point idx).
message MapRequest {
    repeated Point centroids = 1;
    repeated int32 point_ids = 2;
    int32 num_reducers = 3;
    string write_mode = 4;
}

// returns the closest centroid ID for the point. 
message MapResponse {
    bool success = 1;
}

message GetPairsRequest {
    int32 reducer_id = 1;
}

message GetPairsResponse {
    repeated int32 keys = 1;
    repeated Point values = 2;
}

message KeyValue {
    int32 closest_centroid = 1;
    Point point = 2;
}

message Point {
    repeated float dim_val = 1;
}